docs/guide/custom-functions.md

```diff
@@ -46,25 +71,79 @@ import { FunctionPlugin } from 'hyperformula';
+| `expandRanges` | Boolean | If set to `true`, ranges in the function's arguments are inlined to (possibly multiple) scalar arguments. |
+| `isDependentOnSheetStructureChange` | Boolean | If set to `true`, the function gets recalculated with each sheet shape change. |
+| `isVolatile` | Boolean | If set to `true`, the function is [volatile](volatile-functions.md). |
+| `repeatLastArgs` | Number | For functions with a variable number of arguments: sets how many last arguments can be repeated indefinitely. |
+| `vectorizationForbidden` | Boolean | If set to `true`, prevents the function from ever being vectorized (but it's up to your function implementation to properly handle vectorization). |
+
+### Argument validation options
+
+In an optional `parameters` object, you can set rules for your function's argument validation.
+
+```javascript
+// import FunctionPlugin
+import { FunctionPlugin } from 'hyperformula';
+
+// start creating a class
+class CountHF extends FunctionPlugin {
+
+// define functions inside this plugin
+  public static implementedFunctions = {
+    'HYPER': {
+      // this method's functionality will be defined below
+      method: 'hyper',
+      // set your argument validation options below
+      parameters: {
+        passSubtype: false,
+        defaultValue: 10,
+        optionalArg: false,
+        minValue: 5,
+        maxValue: 15,
+        lessThan: 15,
+        greaterThan: 5
+      },
     }
   };
 }
 ```
 
-Similarly, there are other useful properties. 
-`isDependentOnSheetStructureChange` marks functions that need to be recalculated with 
-each change of the shape of the engine sheets. 
-`doesNotNeedArgumentsToBeComputed` marks functions that treat references or ranges in their arguments
-as arguments that do not create dependency. Other arguments are properly evaluated.
-`arrayFunction` denotes functions that enable array arithmetic in its arguments and nested expressions.
-`vectorizationForbidden` when set prevents function from ever being vectorized (however, it is up to implementation of a function to properly handle vectorization).
+You can set the following argument validation options:
+
+| Parameter | Type | Description |
+| --- | --- | --- |
+| `passSubtype` | Boolean | If set to `true`, arguments need to be passed with full type information.<br>(e.g. for numbers: `Date` or `DateTime` or `Time` or `Currency` or `Percentage`) |
+| `defaultValue` | `InternalScalarValue` \| `RawScalarValue` | If an argument is missing, its value defaults to `defaultValue`. |
+| `optionalArg` | Boolean | If set to `true`: if an argument is missing, and no `defaultValue` is set, the argument defaults to `undefined` (instead of throwing an error).<br><br>This is logically equivalent to setting `defaultValue` to `undefined`. |
```

Comment:
Those are interesting subject for a subsection. Why we have both options? Can we have optionalArg between required ones? Should arguments in the middle have default value then? Can we run function with empty args like `=LPV(1, , , 12)`?

Reply:
@izulin can you help with the above questions? Thanks!

Reply:
both options are to handle varying number of args:

setting `defaultValue` to any value makes the argument `optionalArg` automatically.
`optionalArg` is there for situations where we really do not want to set `defaultValue` (the latter automatically supplies defaultValue to the function, so the function is not aware of actual number of true arguments provided), e.g. function has two different behaviours depending on number of arguments

of course those flags are from `runFunction` wrapper

Reply:
> Can we have optionalArg between required ones? 

yes, but that would give no effect

> Should arguments in the middle have default value then? 
> Can we run function with empty args like `=LPV(1, , , 12)`?

No, this would result in `=LPV(1, null, null, 12)` with `null` being the internal `emptyValue`, which will be coerced to `0`, `""` or any other equivalent of `null`.

Reply:
@wojciechczerniak - done: e916b01